### PR TITLE
[SDEV-1827] undo redo shape drawing action

### PR DESCRIPTION
**Pull request by nandishjpatel**
NOTE: https://github.com/delmic/odemis/pull/2701 must be reviewed and merged first

### CODE REVIEW

File: src/odemis/gui/comp/overlay/shapes.py [hunk 1]
```diff
@@ -232,6 +273,12 @@ def on_char(self, evt):
                 self._shape_to_copy = self._selected_shape
                 self.cnvs.set_default_cursor(wx.CURSOR_BULLSEYE)
                 self.cnvs.request_drawing_update()
+            elif evt.GetKeyCode() == wx.WXK_CONTROL_Z:
+                self._selected_shape.undo()
+                self._selected_shape.refresh()
+            elif evt.GetKeyCode() == wx.WXK_CONTROL_Y:
+                self._selected_shape.redo()
+                self._selected_shape.refresh()
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:281]: That mean the undo/redo stack is per *shape*. Isn't that really odd for the user? Typically programs provide a single undo stack. Did you check with the users that it's what they want? I guess that once you're aware of it, it fulfil its need, but the fact that as soon as you select another shape you cannot undo the changes you've made by pressing Ctrl+Z is confusing.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 2]
```diff
@@ -45,6 +47,11 @@ def __init__(self, cnvs):
         # Useful for internal points manipulation
         self._points: List[Vec] = []
         self.cnvs = cnvs
+        # History of states
+        self.undo_stack = deque()
+        self.redo_stack = deque()
+        self.undo_limit = 25
+        self.redo_limit = 25
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:54]: Please put a constant at the top of the file `UNDO_STACK_DEPTH = 25`

File: src/odemis/gui/comp/overlay/ellipse.py [hunk 3]
```diff
@@ -74,10 +84,36 @@ def move_to(self, pos):
...
         self.points.value = self._points
 
+    def save_state(self):
+        """Save the current state of the shape."""
+        state = {
+            EllipseStateKeys.P_POINT1: self.p_point1,
+            EllipseStateKeys.P_POINT2: self.p_point2,
+            EllipseStateKeys.P_POINT3: self.p_point3,
+            EllipseStateKeys.P_POINT4: self.p_point4,
+            EllipseStateKeys.POINTS_PRIVATE: self._points.copy()
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/ellipse.py:98]: `._points` is computed based on the 4 p_points, so I'd argue it doesn't need to be stored as a state.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 4]
```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
...
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No redo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    @abstractmethod
+    def save_state(self):
+        """Save the current state of the shape."""
+        pass
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:132]: I think you should be able to simplify a little bit the code by having a simpler `get_state` function that just returns the state of the shape. Then `save_state()` could be generic here, calling `get_state()` and appending it to the the redo stack (with a depth limit).

It would also allow to simplify the .copy(s) function to just `self.restore_state(s.get_state())`.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 5]
```diff
@@ -232,6 +273,12 @@ def on_char(self, evt):
                 self._shape_to_copy = self._selected_shape
                 self.cnvs.set_default_cursor(wx.CURSOR_BULLSEYE)
                 self.cnvs.request_drawing_update()
+            elif evt.GetKeyCode() == wx.WXK_CONTROL_Z:
+                self._selected_shape.undo()
+                self._selected_shape.refresh()
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:278]: I'd argue the `refresh()` should be done automatically by the `.restore_state()`.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 6]
```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
...
+
+    def redo(self):
+        """Redo the last undone action."""
+        if self.redo_stack:
+            state = self.redo_stack.pop()
+            self.undo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:125]: "warning" means that the error log panel will turn orange. Just for Ctrl+Z or Ctrl+Y, that sounds a little over the top. I'd argue a "info" level is enough.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 7]
```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
...
-        """Refresh the shape's points VA after copy or move_to operation."""
+        """Refresh the shape's points VA after copy, move_to or restore_state operation."""
+        pass
+
+    def undo(self):
+        """Undo the last action."""
+        if self.undo_stack:
+            state = self.undo_stack.pop()
+            self.redo_stack.append(state)
+            self.restore_state(state)
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:110]: Does this work properly? Isn't there some sort of "off by 1" error?
The current state is the "topest" state on the undo_stack, right? Then the first time you call it, nothing will be undone. 
I think you should restore the state to `self.undo_stack[-1]`

File: src/odemis/gui/comp/overlay/shapes.py [hunk 8]
```diff
@@ -45,6 +47,11 @@ def __init__(self, cnvs):
         # Useful for internal points manipulation
         self._points: List[Vec] = []
         self.cnvs = cnvs
+        # History of states
+        self.undo_stack = deque()
+        self.redo_stack = deque()
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:52]: If you use deque, then you should `max_len` to automatically discard the oldest elements. Otherwise, using a list is as simple.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 9]
```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
...
+            self.redo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:118]: Shouldn't the redo stack be automatically emptied when `save_state` is called? Admittedly, I think that it works fine as it is, but my impression is that most software have this behaviour, so it would avoid "surprises" for the user.

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/comp/overlay/shapes.py (hunk 1 above)
Comment (nandishjpatel) [reply src/odemis/gui/comp/overlay/shapes.py:281]: Right, I will change the implementation to `Typically programs provide a single undo stack.`

File: src/odemis/gui/comp/overlay/shapes.py (hunk 9 above)
Comment (nandishjpatel) [reply src/odemis/gui/comp/overlay/shapes.py:118]: Good catch. I had this implemented and for some reason its not here in the PR.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 10]
```diff
@@ -20,13 +20,18 @@
...
+from enum import IntEnum
+import logging
+from typing import Tuple, List, Union, Dict, Any
 
 import wx
 
 from odemis import model, util
 from odemis.gui.comp.overlay.base import WorldOverlay, Vec
 
+UNDO_STACK_DEPTH = 25
```
Comment (tepals) [inline src/odemis/gui/comp/overlay/shapes.py:33]: can you add a short comment to explain what this value is? Something like: `the number of undo actions stored in the stack`

File: src/odemis/gui/comp/overlay/shapes.py [hunk 11]
```diff
@@ -211,7 +232,13 @@ def on_char(self, evt):
         if not self.active.value:
             return super().on_char(evt)
 
-        if self._selected_shape:
+        if evt.GetKeyCode() == wx.WXK_CONTROL_Z:
+            self._undo_action = True
+            self.undo()
+        elif evt.GetKeyCode() == wx.WXK_CONTROL_Y:
```
Comment (tepals) [inline src/odemis/gui/comp/overlay/shapes.py:238]: A lot of software (at least the I use) supports both `ctrl+y` and `ctrl+shift+z`, I find `ctrl+shift+z` easier to use, would it be easy to also support that here?

Comment (nandishjpatel) [reply src/odemis/gui/comp/overlay/shapes.py:238]: Its not that easy
https://docs.wxpython.org/wx.KeyCode.enumeration.html#wx-keycode

**Comment by tepals** [comment]
I tested the PR locally and I noticed that the creation of a shape is not undone when pressing `ctrl+z`. However the action is added to the stack. If I move a shape and then create two new shapes, I need to press `ctrl+z` three times before the movement action is undone. I would expect that either the movement action is undone with one press of `ctrl+z`, or that the creation of shapes is also undone. 


https://github.com/delmic/odemis/assets/46314774/e4ab21c6-ca28-462b-8f1b-34f3c1c47625

**Comment by nandishjpatel** [comment]
> I tested the PR locally and I noticed that the creation of a shape is not undone when pressing `ctrl+z`. However the action is added to the stack. If I move a shape and then create two new shapes, I need to press `ctrl+z` three times before the movement action is undone. I would expect that either the movement action is undone with one press of `ctrl+z`, or that the creation of shapes is also undone.
> undo-redo.mp4

The creation of a shape and copying of a shape is also handled by the undo and redo action now.

**Comment by tepals** [comment]
Should we also handle deleting shapes? If I delete a shape, nothing happens when I press `ctrl+z`

https://github.com/delmic/odemis/assets/46314774/d7a7c1d8-4076-495e-bd94-57c077054761

**Comment by nandishjpatel** [comment]
> Should we also handle deleting shapes? If I delete a shape, nothing happens when I press `ctrl+z`
> Screencast.2024-05-22.17.41.13.mp4

Yeah we should, I forgot about it. By the way I could not play any of your uploaded videos, it says the file is corrupt.

**Comment by nandishjpatel** [comment]
@tepals a suggestion, you should actually test the undo and redo in the software lag fix PR https://github.com/delmic/odemis/pull/2752. The algorithm works better compared this PR because there we let ShapesOverlay handle all mouse events. Just give it a try.

**Comment by nandishjpatel** [comment]
The shape deletion is also handled by the undo and redo action now

File: src/odemis/gui/comp/overlay/shapes.py [hunk 12]
```diff
@@ -237,9 +288,73 @@ def on_left_up(self, evt):
...
+            logging.info(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
+        if self._redo_stack:
+            # If the undo stack is empty or if the previous action was undo
+            # correct the off by 1 shape's state by popping it and not restoring the state
+            if not self._undo_stack or self._undo_action:
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:334]: I don't think the `not undo_stack` is useful here. That's because undo/redo are not symmetrical: editing a shape reset just the redo_stack. So the only way to have an empty undo_stage is that the last action was an undo_action. Hence, it's not useful here.

File: src/odemis/gui/comp/overlay/shapes.py [hunk 13]
```diff
@@ -237,9 +288,73 @@ def on_left_up(self, evt):
...
+                shape_state = ShapeState(self._selected_shape, state, self._is_new_shape, False)
+                if not self._undo_stack or self._undo_stack[-1] != shape_state:
+                    self._undo_stack.append(shape_state)
+                    self._redo_stack.clear()  # Clear redo stack when a shape's state is saved
         else:
             WorldOverlay.on_left_up(self, evt)
 
+    def undo(self):
+        """Undo the last action."""
+        if self._undo_stack:
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/shapes.py:302]: You could make the code  a tiny simpler to read by adjusting this `if` to
```
if not self._undo_stack:
     logging.warning(...)
     return
```
Then, the rest of the code is one indentation less, and become the "normal code".

File: src/odemis/gui/comp/overlay/polygon.py [hunk 14]
```diff
@@ -71,6 +75,25 @@ def move_to(self, pos):
...
+            PolygonStateKeys.FINISHED_PRIVATE: self._finished,
+            PolygonStateKeys.POINTS_PRIVATE: self._points.copy(),
+        }
+        # Only return the state if the polygon creation is finished
+        # By doing so avoid storing an undo action during polygon creation
+        return state if self.right_click_finished else {}
+
+    def restore_state(self, state: Dict[PolygonStateKeys, Any]):
+        """Restore the shape to a given state."""
+        self.v_points = state[PolygonStateKeys.V_POINTS]
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/polygon.py:91]: `v_points` are just the `_points` in the current view coordinate. They are going to be recomputed based on _points in the call to `_phys_to_view()`. So what's the point to store them? I'd suggest to not store them, and just reset v_points to `[None] * len(_points)` to make `_phys_to_view()` happy.

File: src/odemis/gui/comp/overlay/ellipse.py [hunk 15]
```diff
@@ -76,6 +79,29 @@ def move_to(self, pos):
...
+        # By doing so avoid storing an undo action during ellipse creation
+        return state if self.p_point1 != self.p_point3 else {}
+
+    def restore_state(self, state: Dict[EllipseStateKeys, Any]):
+        """Restore the shape to a given state."""
+        self.p_point1 = state[EllipseStateKeys.P_POINT1]
+        self.p_point2 = state[EllipseStateKeys.P_POINT2]
+        self.p_point3 = state[EllipseStateKeys.P_POINT3]
+        self.p_point4 = state[EllipseStateKeys.P_POINT4]
+        self._points = state[EllipseStateKeys.POINTS_PRIVATE]
```
Comment (pieleric) [inline src/odemis/gui/comp/overlay/ellipse.py:101]: The first thing that happens in `draw()` is `self._points.clear()`. They are dependent only on the 4 p_points. So what's the point of storing them? Is it because storing them is easier than recomputing them? If so, please add a comment explaining this. Otherwise, don't store them, and force the recomputation.

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/comp/overlay/ellipse.py (hunk 15 above)
Comment (nandishjpatel) [reply src/odemis/gui/comp/overlay/ellipse.py:101]: I store them, just to set the `points` VA in line 103. Settings the points VA is important because it does the megafield computation in the subscribers callback. If I don't store them, I have to call `update_drawing()` first and then I can set the `points` VA. Doing this will call `update_drawing` twice, once here and then in ShapesOverlay's `undo` or `redo`

File: src/odemis/gui/comp/overlay/rectangle.py [hunk 16]
```diff
@@ -66,6 +70,28 @@ def move_to(self, pos):
...
+        # Only return the state if the rectangle creation is finished
+        # By doing so avoid storing an undo action during rectangle creation
+        return state if self.p_point1 != self.p_point3 else {}
+
+    def restore_state(self, state: Dict[RectangleStateKeys, Any]):
+        """Restore the shape to a given state."""
+        self.p_point1 = state[RectangleStateKeys.P_POINT1]
+        self.p_point2 = state[RectangleStateKeys.P_POINT2]
+        self.p_point3 = state[RectangleStateKeys.P_POINT3]
+        self.p_point4 = state[RectangleStateKeys.P_POINT4]
```
[the remark this answers is not in the document]
Comment (nandishjpatel) [reply src/odemis/gui/comp/overlay/rectangle.py:90]: Right, I changed it

**Review by pieleric — CHANGES_REQUESTED** [review]

**Review by tepals — APPROVED** [review]
